src/wire/icmpv6.rs

```diff
@@ -0,0 +1,690 @@
+            &DstUnreachable::BeyondScope =>
+                write!(f, "beyond scope of source address"),
+            &DstUnreachable::AddrUnreachable =>
+                write!(f, "address unreachable"),
+            &DstUnreachable::PortUnreachable =>
+                write!(f, "port unreachable"),
+            &DstUnreachable::FailedPolicy =>
+                write!(f, "source address failed ingress/egress policy"),
+            &DstUnreachable::RejectRoute =>
+                write!(f, "reject route to destination"),
+            &DstUnreachable::Unknown(id) =>
+                write!(f, "{}", id)
+        }
+    }
+}
+
+/// A read/write wrapper around an Internet Control Message Protocol version 6 packet buffer.
+#[derive(Debug, PartialEq)]
+pub struct Packet<T: AsRef<[u8]>> {
+    buffer: T
+}
+
+// Ranges and constants describing key boundaries in the ICMPv6 header.
+// See https://tools.ietf.org/html/rfc4443 for details.
+mod field {
+    use wire::field::*;
+
+    pub const TYPE:       usize = 0;
+    pub const CODE:       usize = 1;
+    pub const CHECKSUM:   Field = 2..4;
+
+    pub const UNUSED:     Field = 4..8;
+    pub const MTU:        Field = 4..8;
+    pub const POINTER:    Field = 4..8;
+    pub const ECHO_IDENT: Field = 4..6;
+    pub const ECHO_SEQNO: Field = 6..8;
+}
+
+impl<T: AsRef<[u8]>> Packet<T> {
+    /// Imbue a raw octet buffer with ICMPv6 packet structure.
+    pub fn new(buffer: T) -> Packet<T> {
+        Packet { buffer }
+    }
+
+    /// Shorthand for a combination of [new] and [check_len].
+    ///
+    /// [new]: #method.new
+    /// [check_len]: #method.check_len
+    pub fn new_checked(buffer: T) -> Result<Packet<T>> {
+        let packet = Self::new(buffer);
+        packet.check_len()?;
+        Ok(packet)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    pub fn check_len(&self) -> Result<()> {
+        let len = self.buffer.as_ref().len();
+        if len < self.header_len() {
+            Err(Error::Truncated)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Consume the packet, returning the underlying buffer.
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the message type field.
+    #[inline]
+    pub fn msg_type(&self) -> Message {
+        let data = self.buffer.as_ref();
+        Message::from(data[field::TYPE])
+    }
+
+    /// Return the message code field.
+    #[inline]
+    pub fn msg_code(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::CODE]
+    }
+
+    /// Return the checksum field.
+    #[inline]
+    pub fn checksum(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u16(&data[field::CHECKSUM])
+    }
+
+    /// Return the identifier field (for echo request and reply packets).
+    ///
+    /// # Panics
```

Comment:
None of these may panic if `check_len` verifies that the packet is bigger than 8 bytes, too.

Reply:
True. Maybe this should be updated to something like

```
# Warning

This function may return invalid data if the packet is...
```

Reply:
That's obviously true from its name. The notice about panics was just documenting an implementation detail, really.